codechecker_common/review_status_handler.py

```diff
@@ -0,0 +1,283 @@
+                return SourceReviewStatus(
+                    status=item['review_status'],
+                    message=item['message']
+                    .encode(encoding='utf-8', errors='ignore')
+                    if 'message' in item else b'',
+                    bug_hash=report.report_hash,
+                    in_source=True)
+
+        return None
+
+    def get_review_status_from_source(
+        self,
+        report: Report
+    ) -> Optional[SourceReviewStatus]:
+        """
+        Return the review status based on the source code comment belonging to
+        the given report.
+
+        - Return the review status if it is set in the source code.
+        - If the review status is ambiguous (i.e. multiple source code
+            comments belong to it) then a ValueError exception is raised which
+            contains information about the problem in a string.
+        - If the soure file changed (which means that the source comments may
+            have replaced, changed or removed) then None returns.
+        - If no review status belongs to the report in the source code, then
+            return None.
+        TODO: This function either returns a SourceReviewStatus, or raises an
+        exception or returns None. This is too many things that a caller needs
+        to handle. The reason is that according to the current behaviors,
+        ambiguous comment results a hard error, and changed files result a
+        warning with "unreviewed" status. In the future we could implement a
+        logic where we take the first comment into account in case of ambiguity
+        or return "unreviewed" with a warning, like changed files.
+        """
+        # The original file path is needed here, not the trimmed, because
+        # the source files are extracted as the original file path.
+        # TODO: Should original_path be strip('/') at store?
+        source_file_name = os.path.realpath(os.path.join(
+            self.__source_root, report.file.original_path))
+
+        if source_file_name in report.changed_files:
+            return None
+
+        if os.path.isfile(source_file_name):
+            src_comment_data = self.__parse_codechecker_review_comment(
+                source_file_name, report.line, report.checker_name)
+
+            if len(src_comment_data) == 1:
+                data = src_comment_data[0]
+                status, message = data.status, data.message
+                self.__source_commets[report] = data
+
+                return SourceReviewStatus(
+                    status=status,
+                    message=message.encode('utf-8'),
+                    bug_hash=report.report_hash,
+                    in_source=True)
+            elif len(src_comment_data) > 1:
+                raise ValueError(
+                    f"Multiple source code comments can be found for "
+                    f"'{report.checker_name}' checker in '{source_file_name}' "
+                    f"at line {report.line}.")
```

Comment:
Why do you use exceptions? Aren't exceptions supposed to be, you know, exceptional (and somewhat unforeseen) errors? The user making a source code suppression spaghetti seems anything but exception-worthy.

Reply:
I couldn't find another way of communicating the type of the error. In an exception object I could describe to the user what is the problem. As a common rule-of-thumb I agree with throwing exceptions on exceptional events. However, according to my experiences in Python different kind of errors are also indicated with exceptions. This is why I chose `ValueError` which tells that the value of some config options is wrong.
Off topic: it would be more elegant in Go language, where it is conventional to return an error message which is handled specifically for communicating errors to the caller :)